policybot/pkg/testresults/testresults.go

```diff
@@ -0,0 +1,336 @@
+		bucketName: bucketName,
+	}, nil
+}
+
+func (prt *PrResultTester) query(prefix string) ([]string, error) {
+	ctx := prt.ctx
+	client := prt.Client
+	bucket := client.Bucket(prt.bucketName)
+	query := &storage.Query{Prefix: prefix, Delimiter: "/"}
+	it := bucket.Objects(ctx, query)
+	paths := []string{}
+	for {
+		attrs, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		paths = append(paths, attrs.Prefix)
+
+	}
+	return paths, nil
+}
+
+/*
+ * GetTest function get all directories under the given pr in istio-prow/pr-logs/pull/istio-istio/PRNUMBER for each test suite name.
+ * Client: client used to get buckets and objects.
+ * PrNum: the PR number inputted.
+ * Return []Tests return a slice of Tests objects.
+ */
+func (prt *PrResultTester) getTests(prNumInt int64) (map[string][]string, error) {
+	prNum := strconv.FormatInt(prNumInt, 10)
+	prefixForPr := "pr-logs/pull/istio_istio/" + prNum + "/"
+	testNames, err := prt.query(prefixForPr)
+	if err != nil {
+		return nil, err
+	}
+	testMap := map[string][]string{}
+	var runs []string
+	for _, testPref := range testNames {
+		testPrefSplit := strings.Split(testPref, "/")
+		testname := testPrefSplit[len(testPrefSplit)-2]
+		runs, err = prt.query(testPref)
+		if err != nil {
+			return nil, err
+		}
+		var runPaths []string
+		var ok bool
+		runPaths, ok = testMap[testname]
+		if !ok {
+			runPaths = []string{}
+		}
+		for _, runPath := range runs {
+			if runPath != "" {
+				runPaths = append(runPaths, runPath)
+			}
+		}
+		testMap[testname] = runPaths
+	}
+	return testMap, nil
+}
+
+func (prt *PrResultTester) getInformationFromFinishedFile(pref string, eachRun *store.TestResult) (*store.TestResult, error) {
+	// It is possible that the folder might not contain finished.json.
+	client := prt.Client
+	bucket := client.Bucket(prt.bucketName)
+	newObj := bucket.Object(pref + "finished.json")
+	nrdr, nerr := newObj.NewReader(prt.ctx)
+	var finish finished
+
+	if nerr != nil {
+		return nil, nerr
+	}
+
+	defer nrdr.Close()
+	finishFile, err := ioutil.ReadAll(nrdr)
+	if err != nil {
+		return nil, nerr
+	}
+	err = json.Unmarshal(finishFile, &finish)
+	if err != nil {
+		return eachRun, err
+	}
+
+	passed := finish.Passed
+	result := finish.Result
+	t := finish.Timestamp
+	tm := time.Unix(t, 0)
+
+	eachRun.TestPassed = passed
+	eachRun.Result = result
+	eachRun.FinishTime = tm
+
+	return eachRun, nil
+}
+
+func (prt *PrResultTester) getInformationFromStartedFile(pref string, eachRun *store.TestResult) (*store.TestResult, error) {
+	client := prt.Client
+	bucket := client.Bucket(prt.bucketName)
+	newObj := bucket.Object(pref + "started.json")
+	nrdr, nerr := newObj.NewReader(prt.ctx)
+	if nerr != nil {
+		return nil, nerr
+	}
+
+	defer nrdr.Close()
+	startFile, nerr := ioutil.ReadAll(nrdr)
+	if nerr != nil {
+		return nil, nerr
+	}
+
+	var started started
+	err := json.Unmarshal(startFile, &started)
+	if err != nil {
+		return eachRun, err
```

Comment:
Should this be "return nil, err"?

Reply:
If the read error is not nil, it means that the file does not exist. So I want to return the original eachRun object without adding new fields to it.